fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,527 @@
+        if (finalResultsEmitted) {
+            return CompletableFuture.completedFuture(RecordCursorResult.exhausted());
+        }
+        return subspaceProvider.getSubspaceAsync(context).thenCompose(subspace -> {
+            KeyValueCursor kvCursor = KeyValueCursor.Builder.withSubspace(subspace).setContext(this.context).setContinuation(this.kvCursorContinuation).setScanProperties(scanProperties).build();
+            return kvCursor.forEachResult(nextKv -> {
+                updateStatistics(nextKv.get());
+            }).handle((resultKv, err) -> {
+                if (err == null) {
+                    //resultKV.hasNext() is false and so determine whether this is because of in-band reason or a continuable out-of-band reason
+                    if (resultKv.getNoNextReason() == NoNextReason.SOURCE_EXHAUSTED) {
+                        finalResultsEmitted = true;
+                        this.nextStatsResult = RecordCursorResult.withNextValue(emitFinalStats(), new SizeStatisticsCollectorCursorContinuation(resultKv));
+                    } else {
+                        //the underlying cursor did not produce a row but there are more, return a continuation and propagate the underlying no next reason
+                        this.nextStatsResult = RecordCursorResult.withoutNextValue(new SizeStatisticsCollectorCursorContinuation(resultKv), resultKv.getNoNextReason());
+                    }
+                    return this.nextStatsResult;
+                } else {
+                    throw new RuntimeException(err);  //might be retriable, caller will decide what to do
+                }
+            }).whenComplete((ignore, err) -> kvCursor.close()); //no need to keep it open as it must be reconstituted with a continuation later
+        });
+    }
+
+    //update partial stats
+    private void updateStatistics(@Nonnull KeyValue kv) {
+        this.sizeStatisticsResults.keyCount += 1;
+        this.sizeStatisticsResults.keySize += kv.getKey().length;
+        this.sizeStatisticsResults.maxKeySize = Math.max(this.sizeStatisticsResults.maxKeySize, kv.getKey().length);
+        this.sizeStatisticsResults.valueSize += kv.getValue().length;
+        this.sizeStatisticsResults.maxValueSize = Math.max(this.sizeStatisticsResults.maxValueSize, kv.getValue().length);
+        int totalSize = kv.getKey().length + kv.getValue().length;
+        if (totalSize > 0) {
+            this.sizeStatisticsResults.sizeBuckets[Integer.SIZE - Integer.numberOfLeadingZeros(totalSize) - 1] += 1;
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return context.getExecutor();
+    }
+
+    @Nonnull
+    @Override
+    public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+        visitor.visitEnter(this);
+        return visitor.visitLeave(this);
+    }
+
+    @Nonnull
+    private SizeStatisticsResults emitFinalStats() {
+        return sizeStatisticsResults;
```

Comment:
Should this set `finalResultsEmitted` to `true`? Or maybe it should be removed?

Reply:
It could. The thinking here is that the final stats could (someday) be different than that of the base stats. E.g. compute some derived stats and cache them in a different structure. Right now these derived stats are happening in the methods associated with the results object.

Reply:
I will remove this method until that day to avoid confusion...